datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs

```diff
@@ -85,44 +85,35 @@ impl SimplifyExpressions {
         };
         let info = SimplifyContext::new(execution_props).with_schema(schema);
 
-        let simplifier = ExprSimplifier::new(info);
-
         let new_inputs = plan
             .inputs()
             .iter()
             .map(|input| Self::optimize_internal(input, execution_props))
             .collect::<Result<Vec<_>>>()?;
 
-        let expr = match plan {
-            // Canonicalize step won't reorder expressions in a Join on clause.
-            // The left and right expressions in a Join on clause are not commutative,
-            // since the order of the columns must match the order of the children.
-            LogicalPlan::Join(_) => {
-                plan.expressions()
-                    .into_iter()
-                    .map(|e| {
-                        // TODO: unify with `rewrite_preserving_name`
-                        let original_name = e.name_for_alias()?;
-                        let new_e = simplifier.simplify(e)?;
-                        new_e.alias_if_changed(original_name)
-                    })
-                    .collect::<Result<Vec<_>>>()?
-            }
-            _ => {
-                plan.expressions()
-                    .into_iter()
-                    .map(|e| {
-                        // TODO: unify with `rewrite_preserving_name`
-                        let original_name = e.name_for_alias()?;
-                        let cano_e = simplifier.canonicalize(e)?;
-                        let new_e = simplifier.simplify(cano_e)?;
-                        new_e.alias_if_changed(original_name)
-                    })
-                    .collect::<Result<Vec<_>>>()?
-            }
+        let simplifier = ExprSimplifier::new(info);
+
+        // The left and right expressions in a Join on clause are not commutative,
+        // since the order of the columns must match the order of the children.
+        // Thus, not reorder expressions in a Join `on` clause while simplifying
```

Comment:
Actually, I just took a look at the join node:

https://github.com/apache/arrow-datafusion/blob/94a6192f6be30b7f6d009bc936a866bf5dcb280c/datafusion/expr/src/logical_plan/plan.rs#L2596-L2599

Was it actually a problem with the `filter` clause and not the `on` clause? :thinking: 

Because the `on` here is a vec of tuples to represent left and right (I was assuming it was a single Expr before)

Reply:
I think it's the problem with `on`.  🤔
In my previous experience, it's a problem when the tuples of `on` keeps as before when optimizing, but the `left` and `right` of `Join` change, so they can't match. 
Not pretty sure.

Reply:
I agree it is not entirely clear what is wrong with canonicalizing the expressions for joins. I'll add a note to the comment explaining that we don't really understand the issue (maybe someone can look into it as a follow on)